fdbserver/BlobWorker.actor.cpp

```diff
@@ -4884,165 +4888,210 @@ ACTOR Future<Void> handleFlushGranuleReq(Reference<BlobWorkerData> self, FlushGr
 
 	auto myGranule = self->granuleMetadata.rangeContaining(req.granuleRange.begin);
 	state Reference<GranuleMetadata> metadata = myGranule.cvalue().activeMetadata;
-	if (req.granuleRange != myGranule.range() || !metadata.isValid() || !metadata->cancelled.canBeSet()) {
+	// if req is a client request (epoch == -1), granule ranges may not line up exactly, but if it's from the blob
+	// manager, they should
+	bool reqInRange =
+	    (req.managerEpoch == -1) ? myGranule.range().contains(req.granuleRange) : req.granuleRange == myGranule.range();
+	if (!reqInRange || !metadata.isValid() || !metadata->cancelled.canBeSet()) {
 		if (BW_DEBUG) {
-			fmt::print("BW {0} cannot flush granule [{1} - {2})\n",
-			           self->id.toString().substr(0, 5),
-			           req.granuleRange.begin.printable(),
-			           req.granuleRange.end.printable());
+			fmt::print(
+			    "BW {0} cannot flush req [{1} - {2}) from BM {3}: granule [{4} - {5}), valid: {6}, cancelled: {7}\n",
+			    self->id.toString().substr(0, 5),
+			    req.granuleRange.begin.printable(),
+			    req.granuleRange.end.printable(),
+			    req.managerEpoch,
+			    myGranule.begin().printable(),
+			    myGranule.end().printable(),
+			    metadata.isValid(),
+			    metadata.isValid() && !metadata->cancelled.canBeSet());
 		}
 		req.reply.sendError(wrong_shard_server());
 		return Void();
 	}
 
-	if (metadata->durableDeltaVersion.get() < req.flushVersion) {
-		try {
+	try {
+		if (BW_DEBUG) {
+			fmt::print("BW {0} flushing granule [{1} - {2}) @ {3}\n",
+			           self->id.toString().substr(0, 5),
+			           req.granuleRange.begin.printable(),
+			           req.granuleRange.end.printable(),
+			           req.flushVersion);
+		}
+		state Promise<Void> granuleCancelled = metadata->cancelled;
+		choose {
+			when(wait(metadata->readable.getFuture())) {}
+			when(wait(granuleCancelled.getFuture())) {
+				if (BW_DEBUG) {
+					fmt::print("BW {0} flush granule [{1} - {2}) cancelled 2\n",
+					           self->id.toString().substr(0, 5),
+					           req.granuleRange.begin.printable(),
+					           req.granuleRange.end.printable());
+				}
+				req.reply.sendError(wrong_shard_server());
+				return Void();
+			}
+		}
+
+		// if delta file is already written after flush version, but that delta file did not compact,
+		// we have to write another one to trigger compaction
+		if (req.compactAfter && metadata->pendingSnapshotVersion < req.flushVersion &&
+		    metadata->pendingDeltaVersion >= req.flushVersion) {
+			CODE_PROBE(true, "Bumping granule force flush version to guarantee re-snapshot");
```

Comment:
this is just an indentation change and moving a couple code blocks around, not sure why github lost its mind on this part of the diff